### ISSUE TITLE
twisted.mail.test.test_imap.TLSTestCase.testStartTLS checks for TLS setup in a bad way

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3819]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3819|
|Type|defect|
|Created|2009-05-04 17:55:49Z|
|Branch|https://github.com/twisted/twisted/tree/better-tls-check-3819|

This test uses `ITLSTransport.startTLS` and then checks to see if it worked by looking for 'TLS' in the name of the class of the transport.  Instead, it should check to see if the transport provides `ISSLTransport`.

I came across this as part of [#10385](https://github.com/twisted/twisted/issues/10385), as the test failed because I didn't implement `startTLS` for iocpreactor in a way which caused 'TLS' to become part of the transport class name.

<details><summary>Searchable metadata</summary>

```
trac-id__3819 3819
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_better_tls_check_3819 branches/better-tls-check-3819
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__tests tests
time__1241459749000000 1241459749000000
changetime__1241550513000000 1241550513000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [26833]) Branching to 'better-tls-check-3819'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [26834]) Change the test to check the interface instead of the class name

refs [#3819](https://github.com/twisted/twisted/issues/3819)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @exarkun|
|-|-|

It was not that easy :). test_startTLS doesn't return self.connected, so it's not testing the failure case. Please return both deferreds in the test, then merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [26853]) Wait for `self.connected` as well as the loopback Deferred; `self.connected` is the Deferred that is actually likely to fail

refs [#3819](https://github.com/twisted/twisted/issues/3819)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [26855]) Merge better-tls-check-3819

Author: exarkun
Reviewer: therve
Fixes: [#3819](https://github.com/twisted/twisted/issues/3819)

Change the test for `IMAP4Client.startTLS` to check an interface provided by
the transport, rather than part of the name of the transport, in order to
verify that TLS has actually been started.  This allows it to pass correctly
when run against an alternate `ITLSTransport` implementation which does not
play tricks with the transport's class, and is generally better since
it is checking something which could actually be a public API.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
